### PR TITLE
Basic ego support for JWT

### PR DESCRIPTION
**Pull request by andricDu**
This is a quick solution for the auth until ego-token-middleware is fixed and made avaiable for use with the lastest version of ego. 

### Changes
- Grabs ego public key and caches it
- uses JWT bearer token auth for post/put endpoints
- new error types (unauthorized, forbidden)
- updated swagger with bearer token auth

### CODE REVIEW

File: src/services/egoTokenService.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import memoize from "memoizee";
+import axios from "axios";
+import jwt from "jsonwebtoken";
+import ms from "ms";
+import { NextFunction, Request, Response, RequestHandler } from "express";
+import { ForbiddenError, UnauthorizedError } from "../utils/errors";
+
+const getKey = memoize(
+    async (egoURL: string) => {
+        const response = await axios.get(urlJoin(egoURL, "oauth/token/public_key"));
```
Comment (blabadi) [inline src/services/egoTokenService.ts]: why the other part of the url is hardcoded ?

File: src/services/egoTokenService.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return response.data;
+    },
+    {
+        maxAge: ms("1h"),
```
Comment (blabadi) [inline src/services/egoTokenService.ts]: cool shit man

File: src/services/egoTokenService.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            const { authorization: authorizationBody }: any = req.body || {};
+
+            const authorization = authorizationHeader || authorizationBody;
+            const bearerToken: string = authorization ? authorization.split(" ")[1] : req.query.key;
+
+            let valid = false;
+            try {
+                valid = !!(bearerToken && (await verifyEgoToken(bearerToken, egoURL)));
+            } catch (e) {
+                valid = false;
```
Comment (blabadi) [inline src/services/egoTokenService.ts]: add debug/info logging for the error ? could be useful detecting suspicious behavior

File: src/config/swagger.json [hunk 4]
```diff
@@ -9,6 +9,13 @@
     "http",
```
Comment (blabadi) [inline src/config/swagger.json]: switch this to swagger 3: https://mermade.org.uk/openapi-converter or create the ticket to do that next sprint please

File: src/services/egoTokenService.ts (hunk 1 above)
Comment (andricDu) [reply src/services/egoTokenService.ts]: good point, ill remove this and chop off one dependency.

File: src/services/egoTokenService.ts (hunk 3 above)
Comment (andricDu) [reply src/services/egoTokenService.ts]: :+1:

File: src/config/swagger.json (hunk 4 above)
Comment (andricDu) [reply src/config/swagger.json]: ill be doing this next sprint but good call.

**Comment by andricDu** [comment]
I need to fix unit tests so that they don't break. Auth/Enabled/Disabled flag would be good.

File: src/services/egoTokenService.ts (hunk 1 above)
Comment (blabadi) [reply src/services/egoTokenService.ts]: you are using 4 libraries in one line and you shamed me for adding yaml2js .. :ok_hand: :joy:

Comment (andricDu) [reply src/services/egoTokenService.ts]: :nauseated_face:

File: src/config/swagger.json (hunk 4 above)
Comment (andricDu) [reply src/config/swagger.json]: https://github.com/overture-stack/lectern/issues/17

**Comment by blabadi** [comment]
> I need to fix unit tests so that they don't break. Auth/Enabled/Disabled flag would be good.

I would add a test to make sure the security logic works as expected, specially if someone fiddles with it later, please add a ticket for that

**Pull request merged**